### ISSUE TITLE
Feature Request: `expand` without fractions

### ISSUE THREAD

**Issue by Yuvix25** [sympy/sympy#23059]
Hi,  
I'm writing a code which uses this (wonderful) library, and I have encountered a situation where I need to use expand, except without expanding fractions (so only multiplications and so). I tried turning off everything in the arguments I pass to `sympy.expand`, but none made the fractions stay together. I think it will be helpful if there will also be an argument like `fractions=True` in `sympy.expand`, to help get full control of it.  
Thanks in advance.

**Comment by oscarbenjamin** [comment]
It's not clear to me what the feature request is for. Can you give a clear example? The open issue is not useful if it isn't clear what is being proposed.

**Comment by Yuvix25** [comment]
@smichr Something like that, except not only for Add, but everything except fractions, just like you can set `mul=True/False`.

**Comment by Yuvix25** [comment]
@oscarbenjamin The output of what smichr showed is what I meant, except I suggest it to be more general so that it would work with all the things that `expand` can work with, but with the exception to control whether I want to expand fractions as well or not.

**Comment by oscarbenjamin** [comment]
I don't understand what you mean by "fractions". Please show a clear example of input and output and explain what you want as the desired output. Define clearly what you mean by "fractions".

**Comment by Yuvix25** [comment]
For example if I have this input:
```python
x * (x^2-2y) - (x + y) / 2
```
And I run the regular expand on it, I will get something like this:
```python
x^3 - 2xy - x/2 - y/2
```
So I want to add an argument, which allows me to control whether do I get the above output, or the bellow one:
```python
x^3 - 2xy - (x + y) / 2
```
Is that clear now?

**Comment by oscarbenjamin** [comment]
> Is that clear now?

That's clearer but not clear. By "fractions" do you mean "explicit rational numbers"?

More generally how do you define what you do and don't want to happen? (I don't see how to generalise from this exampkle with the explanation given.)

Note that it is not necessarily expand that does this:
```python
In [1]: x * (x ** 2 - 2 * y) - (x + y) / 2
Out[1]: 
  ⎛ 2      ⎞   x   y
x⋅⎝x  - 2⋅y⎠ - ─ - ─
               2   2
```
Here I haven't used expand but the rational still distributed over the add.

**Comment by Yuvix25** [comment]
Correct, but I used `together` to avoid a case like that (because it joins the rationals), so I want an option within `expand` which lets me keep the rationals together.

**Comment by Yuvix25** [comment]
The general idea is that simply if there's a rational, expand will not break it, so it will stay as one rational (things can happen in the upper part of it, but they must stay within there).

**Comment by oscarbenjamin** [comment]
I still don't understand what you are asking for.

**Comment by Yuvix25** [comment]
Do you agree that I can set `mul=True` or `mul=False` within expand, and it control the output like so:
input:
```python
simple_exp = x * (x+y) + sympy.together((x-y)/2)
```

**Comment by Yuvix25** [comment]
Do you agree that I can set `mul=True` or `mul=False` within expand, and it control the output like so:
input:
```python
simple_exp = x * (x+y) + sympy.together((x-y)/2)
print(sympy.expand(simple_exp, mul=False))
```
Output:
```python
x*(x + y) + (x - y)/2
```
Input 2:
```python
simple_exp = x * (x+y) + sympy.together((x-y)/2)
print(sympy.expand(simple_exp, mul=True))
```
Output 2:
```python
x**2 + x*y + x/2 - y/2
```
?

**Comment by Yuvix25** [comment]
So what I want is to be able to do something like this:
Input:
```python
simple_exp = x * (x+y) + sympy.together((x-y)/2)
sympy.expand(simple_exp, mul=True, rational=False)
```
Output:
```python
x**2 + x*y + (x - y)/2
```

**Comment by oscarbenjamin** [comment]
I don't understand how to generalise what you are asking for to a more complicated situation. What should this do:
```
expand((x + 1) * x + (x - y) / 2)
```
Do you want `x**2 + x + (x - y)/2` or `x**2 + (3*x - y)/2` or something else?

Maybe rather than changing expand what is needed is a way to change the output after calling expand.

**Comment by Yuvix25** [comment]
The first output.

**Comment by Yuvix25** [comment]
And for a more complicated input, if I get this:
```python
(x + 1) * x + (x - y * (1+z)) / 2
```
I would like to get the following output:
```python
x**2 + x + (x - y - y*z)/2
```

**Comment by oscarbenjamin** [comment]
It's quite difficult to preserve a rational Mul because of the 2-arg Mul. I don't know if there is an issue for that. Otherwise I think what you are asking for is quite close to this:
```
In [6]: expr = (x + 1) * x + (x - y * (1 + z)) / 2

In [7]: expr.replace(
   ...:     lambda e: e.is_Mul
   ...:     and len(e.args) == 2
   ...:     and (e.args[0].is_Symbol or e.args[1].is_Symbol),
   ...:     expand_mul,
   ...: )
Out[7]: 
 2   3⋅x   y⋅(z + 1)
x  + ─── - ─────────
      2        2
```
It's hard to say though if that's what you want.

**Comment by oscarbenjamin** [comment]
So far the feature request here does not seem to be well defined. I don't think that it is good to keep an issue open if it isn't understood what the problem is or how it could be fixed.

**Comment by smichr** [comment]
There are a few aspects to this objective:
1) getting terms collected over a common denominator (`together` or `factor_terms` is good for this if you can target the subexpression containing the common denominator)
2) getting the collected terms to stay together when being used in a subexpression (`UnevaluatedExpr` can help with this)
3) traversing an expression with `expand` and only acting on the numerator of `UnevaluatedExpr`

It would be easiest to just expand the expression and then gather terms over common denominators, but you said you didn't want that.

So assuming you can do 1, if you put them into a 2-arg function `M(n, d)` then you can use expand and finally replace `M` with an `UnevaluatedExpr(Mul(1/d, n, evaluate=False))`. Something like this can get you started:
```python
def Mify(e):
        if e.is_Mul:
            n, d = factor_terms(e).as_numer_denom()
            if d.is_Integer and d != 1:
                return M(n, d)
            return e
        if not e.is_Add:
            return e
        # collect common dens in Add
        den = {}
        args = []
        for i in e.args:
            m = factor_terms(i)
            n, d = m.as_numer_denom()
            if d == 1:
                args.append(n)
                continue
            c,d = d.as_coeff_Mul()
            if c not in den:
                den[c] = []
            den[c].append(n/d)
        for k, v in den.items():
            if len(v) == 1:
                [a] = v
                if a.is_Mul and not any(i.is_Add for i in a.args):
                    args.append(a/k)
                else:
                    args.append(M(a, k))
            else:
                args.append(M(Add(*v), k))
        return Add(*args)

def nexpand(eq):
    rv = Mify(eq).expand()
    return rv.replace(M,
        lambda n, d: UnevaluatedExpr(Mul(1/d, n, evaluate=False)))

var('x y z')
e1 = (x + 1) * x + (x - y * (1 + z)) / 2
print('e1 =',e1)
print(Mify(e1))
print(nexpand(e1))
print()
e2 = x*(x+1)+(x+1)/2*x
print('e2 =',e2)
print(Mify(e2))
print(nexpand(e2))
```
output:
```
e1 = x*(x + 1) + x/2 - y*(z + 1)/2
x*(x + 1) + M(x - y*(z + 1), 2)
x**2 + x + (x - y*z - y)/2

e2 = x*(x/2 + 1/2) + x*(x + 1)
x*(x + 1) + M(x*(x + 1), 2)
x**2 + x + (x**2 + x)/2
```

**Comment by smichr** [comment]
> cool if this could be added as a feature

I'm not sure of the use case. 

I can see wanting to collect over a common rational denominator where `2` replaces `x` in 
```python
>>> collect(1/x+y/x+y,x)
y + (y + 1)/x
```
I could also see making the `frac`,`numer` and `denom` apply to subexpressions so `eq + 2` would behave as the following except each would have an additive `2`:
```python
>>> eq=x*(x+1)/(z*(z+1))
>>> eq.expand(frac=True)
(x**2 + x)/(z**2 + z)
>>> eq.expand(numer=True)
(x**2 + x)/(z*(z + 1))
>>> eq.expand(denom=True)
x*(x + 1)/(z**2 + z)
```

**Comment by smichr** [comment]
In `Expr.expand` perhaps this
```python
        expr = self
        if hints.pop('frac', False):
            n, d = [a.expand(deep=deep, modulus=modulus, **hints)
                    for a in fraction(self)]
            return n/d
```
could be changed to
```python
        expr = self
        if hints.pop('frac', False):
            if deep and self.args and not isinstance(self, Mul):
			    self = self.func(*(i.expand(deep=deep,
					modulus=modulus, **hints) for i in self.args])
			n, d = [a.expand(deep=deep, modulus=modulus, **hints)
					for a in fraction(self)]
			return n/d
```
